Add a minor `persistent-scratch' mode

PierreTechoueyres opened the pull request:
This allow to add a mode map and remap `save-buffer' and `write-file'
to the persistent equivalents.  Also add a message when saving.

* persistent-scratch.el (persistent-scratch-save): add a message when called interactively,
                        (persistent-scratch-minor-mode) : new minor mode.

PierreTechoueyres committed:
Add a minor `persistent-scratch' mode

This allow to add a mode map and remap `save-buffer' and `write-file'
to the persistent equivalents.  Also add a message when saving.

* persistent-scratch.el (persistent-scratch-save): add a message when
called interactively,
(persistent-scratch-minor-mode) : new minor mode.

Fanael commented on persistent-scratch.el:
```diff
@@ -154,7 +154,8 @@ representing the time of the last `persistent-scratch-new-backup' call."
           (let ((coding-system-for-write 'utf-8-unix))
             (write-region str nil tmp-file nil 0))
         (set-default-file-modes old-umask)))
-    (rename-file tmp-file actual-file t))
```
This line is absolutely necessary for persistent-scratch to work: it relies on the write-new-then-rename idiom to atomically replace old saved state with new. (It's not a perfect solution because it loses inode attributes like permissions and xattrs, but that's an unix API deficiency that can't be fixed without introducing new APIs for something like atomic extent swap.) Without it, the new state is only saved to the temporary file, which is never read by anything.

PierreTechoueyres replied:
My bad ! Stupid copy / paste error from my init file ...
Should be fixed with new patch.

Fanael commented on persistent-scratch.el:
```diff
@@ -154,7 +154,8 @@ representing the time of the last `persistent-scratch-new-backup' call."
             (write-region str nil tmp-file nil 0))
         (set-default-file-modes old-umask)))
-    (rename-file tmp-file actual-file t))
+    (when (interactive-p)
```
`interactive-p` is obsolete, should be replaced with `called-interactively-p`.

PierreTechoueyres replied:
Should be fixed with new patch.

Fanael reviewed (changes requested):
This is a nice idea, but I'm not sure what's the intended use of `persistent-scratch-mode` here, could you clarify? Is it supposed to be called directly by the user on the buffers they want this behavior in? Is it really meant to ignore `persistent-scratch-scratch-buffer-p-function` altogether? I'd consider raising an error in `persistent-scratch-mode` if it's being enabled in a buffer that's not considered a scratch buffer by that (user-defined!) predicate.

PierreTechoueyres committed:
Fix previous commit

* persistent-scratch.el (persistent-scratch-save): reintroduce
`rename-file' (mistakenly deleted), use `called-interactively-p'
instead of the obsolete `interactive-p'.

PierreTechoueyres commented:
> This is a nice idea, but I'm not sure what's the intended use of `persistent-scratch-mode` here, could you clarify? Is it supposed to be called directly by the user on the buffers they want this behavior in? Is it really meant to ignore `persistent-scratch-scratch-buffer-p-function` altogether? I'd consider raising an error in `persistent-scratch-mode` if it's being enabled in a buffer that's not considered a scratch buffer by that (user-defined!) predicate.

Yes `persistent-scratch-mode` should be called by user. 
 I use it like this in my init.el:
``` elisp
(with-current-buffer "*scratch*"
  (emacs-lock-mode 'kill)
  (persistent-scratch-mode 1)
  (hs-minor-mode 1)
  (hs-hide-all))
```
The main feature is, from my point of view, that I then can save the buffer at will.

For `persistent-scratch-scratch-buffer-p-function` I didn't notice it and maybe it should be checked before enabling the mode. Does the code below is what you mean ?
``` elisp
(define-minor-mode persistent-scratch-mode
  "Preserve the state of scratch buffers.

Preserve the state of scratch buffers across Emacs sessions by saving the state
to and restoring it from a file."
  :lighter " PS"
  (when persistent-scratch-mode
    (unless (funcall persistent-scratch-scratch-buffer-p-function)
      (setq  persistent-scratch-mode nil)
      (user-error "This buffer isn't managed by `persistent-scratch' package. Mode isn't enabled"))))
```
If, so I could amend the patch.

Fanael commented:
Yes, that's what I have in mind. If you do that, I'll gladly merge your patch.

(Sorry for late response, I forgot…)

PierreTechoueyres committed:
Conditionnaly enable `persistent-scratch-mode'

* persistent-scratch-mode.el(persistent-scratch-mode): Only enable
  mode if buffer meet requirements of
  `persistent-scratch-scratch-buffer-p-function'.

PierreTechoueyres commented:
@Fanael: Hello

> Yes, that's what I have in mind. If you do that, I'll gladly merge your patch.

Should be done now.

>  (Sorry for late response, I forgot…)

No worries.

Fanael commented:
Squashed with minor formatting and grammar changes into 636c031a5bc0fc36f2e48e1f39c32eb4f4fd5dff and merged, thanks!
